pkg/controller/podlocality/controller.go

```diff
@@ -0,0 +1,189 @@
+					})
+				}
+			}
+			return requests
+		}),
+	}, predicate.Funcs{
+		CreateFunc: func(evt event.CreateEvent) bool { return evt.Meta != nil },
+		DeleteFunc: func(evt event.DeleteEvent) bool { return false },
+		UpdateFunc: func(evt event.UpdateEvent) bool {
+			return evt.MetaOld != nil && evt.MetaNew != nil &&
+				(evt.MetaOld.GetLabels()[NodeRegionLabel] != evt.MetaNew.GetLabels()[NodeRegionLabel] ||
+					evt.MetaOld.GetLabels()[NodeZoneLabel] != evt.MetaNew.GetLabels()[NodeZoneLabel])
+		},
+		GenericFunc: func(evt event.GenericEvent) bool { return false },
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func hasNode(pod runtime.Object) bool {
+	return pod != nil && pod.(*v1.Pod).Spec.NodeName != ""
+}
+
+func hasSidecarAnnotation(pod runtime.Object) bool {
+	return pod != nil && podHasSidecar(*pod.(*v1.Pod))
+}
+
+func podHasSidecar(pod v1.Pod) bool {
+	return pod.Annotations[IstioSidecarStatusAnnotation] != ""
+}
+
+var _ reconcile.Reconciler = &PodLocalityReconciler{}
+
+// PodLocalityReconciler copies the node's region/zone labels to the pod after it's scheduled to a node
+type PodLocalityReconciler struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	common.ResourceManager
+	scheme *runtime.Scheme
+}
+
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *PodLocalityReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Processing Pod")
+
+	// Fetch the Pod
+	pod := &v1.Pod{}
+	err := r.Client.Get(context.TODO(), request.NamespacedName, pod)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object
+		return reconcile.Result{}, err
+	}
+
+	if pod.Spec.NodeName == "" {
+		// nothing to do. Pod not scheduled yet.
+		return reconcile.Result{}, nil
+	}
+
+	node := &v1.Node{}
+	err = r.Client.Get(context.TODO(), client.ObjectKey{Name: pod.Spec.NodeName}, node)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if node.Labels[NodeRegionLabel] == "" && node.Labels[NodeZoneLabel] == "" {
+		reqLogger.Info("The node the pod is scheduled on neither has the region nor zone labels. Nothing to do.")
+		return reconcile.Result{}, nil
+	}
+
+	if pod.Labels[NodeRegionLabel] == node.Labels[NodeRegionLabel] && pod.Labels[NodeZoneLabel] == node.Labels[NodeZoneLabel] {
+		reqLogger.Info("Pod's locality labels match the node's. Nothing to do.")
+		return reconcile.Result{}, nil
+	}
+
+	pod.Labels[NodeRegionLabel] = node.Labels[NodeRegionLabel]
+	pod.Labels[NodeZoneLabel] = node.Labels[NodeZoneLabel]
```

Comment:
@rcernich Do you think I should add these to the pod as annotations instead of labels?

Reply:
good question.  they'd be hidden a bit more as annotations, but labels are easier to use in selectors.  i don't know.

labels do have limitations on values, but if these are coming from other labels, it's not really a concern.